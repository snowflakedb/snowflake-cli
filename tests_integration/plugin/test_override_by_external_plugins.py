# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import pytest


@pytest.mark.parametrize("config_mode", ["legacy", "config_ng"], indirect=True)
@pytest.mark.integration
def test_override_build_in_commands(
    runner, test_root_path, _install_plugin, caplog, snapshot, config_mode
):
    """Test plugin override attempt with both legacy and config_ng systems."""
    config_path = (
        test_root_path / "config" / "plugin_tests" / "override_plugin_config.toml"
    )

    result = runner.invoke(
        ["--config-file", config_path, "connection", "list", "--format", "JSON"]
    )

    assert (
        "Cannot register plugin [override]: Cannot add command [snow connection list] because it already exists."
        in caplog.messages
    )

    # Use snapshot to capture the output
    # Each config_mode gets its own snapshot automatically
    assert result.output == snapshot


@pytest.mark.parametrize("config_mode", ["legacy", "config_ng"], indirect=True)
@pytest.mark.integration
def test_disabled_plugin_is_not_executed(
    runner, test_root_path, _install_plugin, caplog, snapshot, config_mode
):
    """Test disabled plugin with both legacy and config_ng systems."""
    config_path = (
        test_root_path
        / "config"
        / "plugin_tests"
        / "disabled_override_plugin_config.toml"
    )

    result = runner.invoke(
        ["--config-file", config_path, "connection", "list", "--format", "JSON"]
    )

    # Use snapshot to capture the output
    # Each config_mode gets its own snapshot automatically
    assert result.output == snapshot


@pytest.fixture(scope="module")
def _install_plugin(test_root_path):
    import subprocess

    path = test_root_path / ".." / "test_external_plugins" / "override_build_in_command"
    subprocess.check_call(["pip", "install", path])
