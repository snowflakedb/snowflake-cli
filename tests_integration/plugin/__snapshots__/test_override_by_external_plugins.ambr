# serializer version: 1
# name: test_disabled_plugin_is_not_executed[config_ng]
  '''
  [
      {
          "connection_name": "test",
          "parameters": {
              "account": "test"
          },
          "is_default": false
      },
      {
          "connection_name": "integration",
          "parameters": {
              "account": "po52878",
              "user": "mraba",
              "host": "po52878.snowflakecomputing.com",
              "database": "SNOWCLI_DB",
              "warehouse": "xsmall",
              "role": "integration_tests",
              "authenticator": "SNOWFLAKE_JWT"
          },
          "is_default": false
      }
  ]
  
  '''
# ---
# name: test_disabled_plugin_is_not_executed[legacy]
  '''
  [
      {
          "connection_name": "test",
          "parameters": {
              "account": "test"
          },
          "is_default": false
      }
  ]
  
  '''
# ---
# name: test_override_build_in_commands[config_ng]
  '''
  Outside command code
  [
      {
          "connection_name": "test",
          "parameters": {
              "account": "test"
          },
          "is_default": false
      },
      {
          "connection_name": "integration",
          "parameters": {
              "account": "po52878",
              "user": "mraba",
              "host": "po52878.snowflakecomputing.com",
              "database": "SNOWCLI_DB",
              "warehouse": "xsmall",
              "role": "integration_tests",
              "authenticator": "SNOWFLAKE_JWT"
          },
          "is_default": false
      }
  ]
  
  '''
# ---
# name: test_override_build_in_commands[legacy]
  '''
  Outside command code
  [
      {
          "connection_name": "test",
          "parameters": {
              "account": "test"
          },
          "is_default": false
      }
  ]
  
  '''
# ---
